tests/alerts_tests.py

```diff
@@ -41,95 +50,204 @@
+            Alert(id=5, crontab="* * * * *", active=False, label="alert_5"),
+            Alert(id=6, crontab="* * * * *", active=False, label="alert_6"),
+        ]
+        observers = [
+            SQLObserver(
+                name="observer_1", sql="SELECT 0", alert_id=1, database_id=database_id
             ),
-            Alert(
-                id=2,
-                label="alert_2",
-                active=True,
-                crontab="*/1 * * * *",
-                sql="SELECT 55",
-                alert_type="email",
-                slice_id=slice_id,
-                recipients="recipient1@superset.com",
-                slack_channel="#test_channel",
+            SQLObserver(
+                name="observer_2",
+                sql="SELECT id FROM test_table WHERE id = -1",
+                alert_id=2,
                 database_id=database_id,
             ),
-            Alert(
-                id=3,
-                label="alert_3",
-                active=False,
-                crontab="*/1 * * * *",
-                sql="UPDATE 55",
-                alert_type="email",
-                slice_id=slice_id,
-                database_id=database_id,
+            SQLObserver(
+                name="observer_3", sql="$%^&", alert_id=3, database_id=database_id
+            ),
+            SQLObserver(
+                name="observer_4", sql="SELECT 55", alert_id=4, database_id=database_id
             ),
-            Alert(id=4, active=False, label="alert_4", database_id=-1),
-            Alert(id=5, active=False, label="alert_5", database_id=database_id),
         ]
 
         db.session.bulk_save_objects(alerts)
-        db.session.commit()
+        db.session.bulk_save_objects(observers)
         yield db.session
 
+        db.session.query(SQLObservation).delete()
+        db.session.query(SQLObserver).delete()
+        db.session.query(Validator).delete()
         db.session.query(AlertLog).delete()
         db.session.query(Alert).delete()
 
 
+def test_alert_observer(setup_database):
+    dbsession = setup_database
+
+    # Test SQLObserver with empty SQL return
+    alert2 = dbsession.query(Alert).filter_by(id=2).one()
+    observe(alert2.id)
+    assert alert2.sql_observers[0].observations[-1].value is None
+
+    # Test SQLObserver with non-empty SQL return
+    alert4 = dbsession.query(Alert).filter_by(id=4).one()
+    observe(alert4.id)
+    assert alert4.sql_observers[0].observations[-1].value == 55
+
+
+def test_alert_error(setup_database):
+    dbsession = setup_database
+
+    # Test error with Observer SQL statement
+    alert3 = dbsession.query(Alert).filter_by(id=3).one()
+    check_alert(alert3.id, alert3.label)
+    assert alert3.logs[-1].state == AlertState.ERROR
+
+    # Test error with alert lacking observer
+    alert5 = dbsession.query(Alert).filter_by(id=5).one()
+    check_alert(alert5.id, alert5.label)
+    assert alert5.logs[-1].state == AlertState.ERROR
```

Comment:
let's add successful alert tests as well

Reply:
i chose to abstract that into the different validator tests since an successful alert is based on its SQLObserver value plus its specific validator